src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/BuildHostProcessManager.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Xml;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+
+internal sealed class BuildHostProcessManager : IAsyncDisposable
+{
+    private readonly string? _binaryLogPath;
+
+    private readonly SemaphoreSlim _gate = new(initialCount: 1);
+    private readonly Dictionary<BuildHostProcessKind, BuildHostProcess> _processes = new();
+
+    public BuildHostProcessManager(string? binaryLogPath = null)
+    {
+        _binaryLogPath = binaryLogPath;
+    }
+
+    public async Task<IBuildHost> GetBuildHostAsync(string projectFilePath, CancellationToken cancellationToken)
+    {
+        var neededBuildHostKind = GetKindForProject(projectFilePath);
+
+        using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+        {
+            if (!_processes.TryGetValue(neededBuildHostKind, out var buildHostProcess))
+            {
+                var process = neededBuildHostKind switch
+                {
+                    BuildHostProcessKind.NetCore => LaunchDotNetCoreBuildHost(),
+                    BuildHostProcessKind.NetFramework => LaunchDotNetFrameworkBuildHost(),
+                    _ => throw ExceptionUtilities.UnexpectedValue(neededBuildHostKind)
+                };
+
+                buildHostProcess = new BuildHostProcess(process);
+                buildHostProcess.Disconnected += BuildHostProcess_Disconnected;
+                _processes.Add(neededBuildHostKind, buildHostProcess);
+            }
+
+            return buildHostProcess.BuildHost;
+        }
+    }
+
+    private void BuildHostProcess_Disconnected(object? sender, EventArgs e)
+    {
+        Contract.ThrowIfNull(sender, $"{nameof(BuildHostProcess)}.{nameof(BuildHostProcess.Disconnected)} was raised with a null sender.");
```

Comment:
Lets log something when the process disconnects and we shut down (e.g. trace)

Reply:
Always or just unexpected cases? Note if we disconnect directly we'll get exceptions thrown in the actual load paths.

Reply:
Unexpected definitely - but even in expected cases can sometimes be useful (if it wasn't actually supposed to be expected or something).

Reply:
Done.